src/librustc_mir/borrow_check/nll/member_constraints.rs

```diff
@@ -0,0 +1,226 @@
+    {
+        // We can re-use most of the original data, just tweaking the
+        // linked list links a bit.
+        //
+        // For example if we had two keys Ra and Rb that both now wind
+        // up mapped to the same key S, we would append the linked
+        // list for Ra onto the end of the linked list for Rb (or vice
+        // versa) -- this basically just requires rewriting the final
+        // link from one list to point at the othe other (see
+        // `append_list`).
+
+        let MemberConstraintSet { first_constraints, mut constraints, choice_regions } = self;
+
+        let mut first_constraints2 = FxHashMap::default();
+        first_constraints2.reserve(first_constraints.len());
+
+        for (r1, start1) in first_constraints {
+            let r2 = map_fn(r1);
+            if let Some(&start2) = first_constraints2.get(&r2) {
+                append_list(&mut constraints, start1, start2);
+            }
+            first_constraints2.insert(r2, start1);
+        }
+
+        MemberConstraintSet {
+            first_constraints: first_constraints2,
+            constraints,
+            choice_regions,
+        }
+    }
+}
+
+impl<'tcx, R> MemberConstraintSet<'tcx, R>
+where
+    R: Copy + Hash + Eq,
+{
+    crate fn all_indices(
+        &self,
+    ) -> impl Iterator<Item = NllMemberConstraintIndex> {
+        self.constraints.indices()
+    }
+
+    /// Iterate down the constraint indices associated with a given
+    /// peek-region.  You can then use `choice_regions` and other
+    /// methods to access data.
+    crate fn indices(
+        &self,
+        member_region_vid: R,
+    ) -> impl Iterator<Item = NllMemberConstraintIndex> + '_ {
+        let mut next = self.first_constraints.get(&member_region_vid).cloned();
+        std::iter::from_fn(move || -> Option<NllMemberConstraintIndex> {
+            if let Some(current) = next {
+                next = self.constraints[current].next_constraint;
+                Some(current)
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Returns the "choice regions" for a given member
+    /// constraint. This is the R1..Rn from a constraint like:
+    ///
+    /// ```
+    /// R0 member of [R1..Rn]
+    /// ```
+    crate fn choice_regions(&self, pci: NllMemberConstraintIndex) -> &[ty::RegionVid] {
+        let NllMemberConstraint { start_index, end_index, .. } = &self.constraints[pci];
+        &self.choice_regions[*start_index..*end_index]
+    }
+}
+
+impl<'tcx, R> Index<NllMemberConstraintIndex> for MemberConstraintSet<'tcx, R>
+where
+    R: Copy + Hash + Eq,
+{
+    type Output = NllMemberConstraint<'tcx>;
+
+    fn index(&self, i: NllMemberConstraintIndex) -> &NllMemberConstraint<'tcx> {
+        &self.constraints[i]
+    }
+}
+
+/// Given a linked list starting at `source_list` and another linked
+/// list starting at `target_list`, modify `target_list` so that it is
+/// followed by `source_list`.
```

Comment:
I don't think there's a requirement that the words from the function name appear in the comment. =) The comment was written to convey:

- what append means, in case you don't know (if you already know, then it clarifies exactly *how* the "appending" takes place);
- the fact that this does its mutation "in place".

I'm basically happy with the comment as is, but let know if you have a specific suggestion.